# **Assignment-3.1**

## **SQL-based data analysis using filtering, aggregation, and basic business queries**

## **1. Load Dataset**
- i have uploaded the superstore csv file as table in databriks

In [0]:
select * from sample_superstore limit 10

Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.96,2,0.0,41.9136
2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs, Rounded Back",731.94,3,0.0,219.582
3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters by Universal,14.62,2,0.0,6.8714
4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.031
5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.368,2,0.2,2.5164
6,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,FUR-FU-10001487,Furniture,Furnishings,"Eldon Expressions Wood and Plastic Desk Accessories, Cherry Wood",48.86,7,0.0,14.1694
7,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.28,4,0.0,1.9656
8,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.152,6,0.2,90.7152
9,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by Samsill,18.504,3,0.2,5.7825
10,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9,5,0.0,34.47


## **2. Explore Table**
- Understand columns and sample data.

### **Analysing the schema of the table**

In [0]:
describe sample_superstore;

col_name,data_type,comment
Row ID,bigint,null
Order ID,string,null
Order Date,date,null
Ship Date,date,null
Ship Mode,string,null
Customer ID,string,null
Customer Name,string,null
Segment,string,null
Country,string,null
City,string,null


### **Printing the Sample data of the tables top 10 rows using limit**

In [0]:
select * from sample_superstore limit 10;

Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.96,2,0.0,41.9136
2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs, Rounded Back",731.94,3,0.0,219.582
3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters by Universal,14.62,2,0.0,6.8714
4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.031
5,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.368,2,0.2,2.5164
6,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,FUR-FU-10001487,Furniture,Furnishings,"Eldon Expressions Wood and Plastic Desk Accessories, Cherry Wood",48.86,7,0.0,14.1694
7,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.28,4,0.0,1.9656
8,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.152,6,0.2,90.7152
9,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by Samsill,18.504,3,0.2,5.7825
10,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9,5,0.0,34.47


- Dataset contains sales, customers, products and profit details.

# **3. WHERE Filters**

### **Region Filter**
- Analyzing a particular region.

In [0]:
select distinct Region from sample_superstore limit 10;

Region
South
West
Central
East


- these are the four available regions

In [0]:
select * from sample_superstore where Region='West' limit 10

Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
3,CA-2016-138688,2016-06-12,2016-06-16,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters by Universal,14.62,2,0.0,6.8714
6,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,FUR-FU-10001487,Furniture,Furnishings,"Eldon Expressions Wood and Plastic Desk Accessories, Cherry Wood",48.86,7,0.0,14.1694
7,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,OFF-AR-10002833,Office Supplies,Art,Newell 322,7.28,4,0.0,1.9656
8,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,TEC-PH-10002275,Technology,Phones,Mitel 5320 IP Phone VoIP phone,907.152,6,0.2,90.7152
9,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,OFF-BI-10003910,Office Supplies,Binders,DXL Angle-View Binders with Locking Rings by Samsill,18.504,3,0.2,5.7825
10,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,OFF-AP-10002892,Office Supplies,Appliances,Belkin F5C206VTEL 6 Outlet Surge,114.9,5,0.0,34.47
11,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,FUR-TA-10001539,Furniture,Tables,Chromcraft Rectangular Conference Tables,1706.184,9,0.2,85.3092
12,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,TEC-PH-10002033,Technology,Phones,Konftel 250 Conference�phone�- Charcoal black,911.424,4,0.2,68.3568
14,CA-2016-161389,2016-12-05,2016-12-10,Standard Class,IM-15070,Irene Maddox,Consumer,United States,Seattle,Washington,98103,West,OFF-BI-10003656,Office Supplies,Binders,Fellowes PB200 Plastic Comb Binding Machine,407.976,3,0.2,132.5922
18,CA-2014-167164,2014-05-13,2014-05-15,Second Class,AG-10270,Alejandro Grove,Consumer,United States,West Jordan,Utah,84084,West,OFF-ST-10000107,Office Supplies,Storage,Fellowes Super Stor/Drawer,55.5,2,0.0,9.99


In [0]:
select count(*) from sample_superstore where Region='West'

count(*)
3203


- there are around `3203` records belongs to `west region`

### **Category Filter**

In [0]:
select distinct Category from sample_superstore limit 10;

Category
Furniture
Office Supplies
Technology


- it contains around 3 categories

In [0]:
select * from sample_superstore where Category='Furniture' limit 10;

Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
1,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.96,2,0.0,41.9136
2,CA-2016-152156,2016-11-08,2016-11-11,Second Class,CG-12520,Claire Gute,Consumer,United States,Henderson,Kentucky,42420,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs, Rounded Back",731.94,3,0.0,219.582
4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.031
6,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,FUR-FU-10001487,Furniture,Furnishings,"Eldon Expressions Wood and Plastic Desk Accessories, Cherry Wood",48.86,7,0.0,14.1694
11,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,FUR-TA-10001539,Furniture,Tables,Chromcraft Rectangular Conference Tables,1706.184,9,0.2,85.3092
24,US-2017-156909,2017-07-16,2017-07-18,Second Class,SF-20065,Sandra Flanagan,Consumer,United States,Philadelphia,Pennsylvania,19140,East,FUR-CH-10002774,Furniture,Chairs,"Global Deluxe Stacking Chair, Gray",71.372,2,0.3,-1.0196
25,CA-2015-106320,2015-09-25,2015-09-30,Standard Class,EB-13870,Emily Burns,Consumer,United States,Orem,Utah,84057,West,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,1044.63,3,0.0,240.2649
28,US-2015-150630,2015-09-17,2015-09-21,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,Pennsylvania,19140,East,FUR-BO-10004834,Furniture,Bookcases,"Riverside Palais Royal Lawyers Bookcase, Royale Cherry Finish",3083.43,7,0.5,-1665.0522
30,US-2015-150630,2015-09-17,2015-09-21,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,Pennsylvania,19140,East,FUR-FU-10004848,Furniture,Furnishings,"Howard Miller 13-3/4"" Diameter Brushed Chrome Round Wall Clock",124.2,3,0.2,15.525
37,CA-2016-117590,2016-12-08,2016-12-10,First Class,GH-14485,Gene Hale,Corporate,United States,Richardson,Texas,75080,Central,FUR-FU-10003664,Furniture,Furnishings,"Electrix Architect's Clamp-On Swing Arm Lamp, Black",190.92,5,0.6,-147.963


In [0]:
select * from sample_superstore where Sales>1000 limit 10;

Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
11,CA-2014-115812,2014-06-09,2014-06-14,Standard Class,BH-11710,Brosina Hoffman,Consumer,United States,Los Angeles,California,90032,West,FUR-TA-10001539,Furniture,Tables,Chromcraft Rectangular Conference Tables,1706.184,9,0.2,85.3092
25,CA-2015-106320,2015-09-25,2015-09-30,Standard Class,EB-13870,Emily Burns,Consumer,United States,Orem,Utah,84057,West,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,1044.63,3,0.0,240.2649
28,US-2015-150630,2015-09-17,2015-09-21,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,Pennsylvania,19140,East,FUR-BO-10004834,Furniture,Bookcases,"Riverside Palais Royal Lawyers Bookcase, Royale Cherry Finish",3083.43,7,0.5,-1665.0522
36,CA-2016-117590,2016-12-08,2016-12-10,First Class,GH-14485,Gene Hale,Corporate,United States,Richardson,Texas,75080,Central,TEC-PH-10004977,Technology,Phones,GE 30524EE4,1097.544,7,0.2,123.4737
55,CA-2016-105816,2016-12-11,2016-12-17,Standard Class,JM-15265,Janet Molinari,Corporate,United States,New York City,New York,10024,East,TEC-PH-10002447,Technology,Phones,AT&T CL83451 4-Handset Telephone,1029.95,5,0.0,298.6855
68,CA-2014-106376,2014-12-05,2014-12-10,Standard Class,BS-11590,Brendan Sweed,Corporate,United States,Gilbert,Arizona,85234,West,OFF-AR-10002671,Office Supplies,Art,"Hunt BOSTON Model 1606 High-Volume Electric Pencil Sharpener, Beige",1113.024,8,0.2,111.3024
150,CA-2016-114489,2016-12-05,2016-12-09,Standard Class,JE-16165,Justin Ellison,Corporate,United States,Franklin,Wisconsin,53132,Central,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs, Rounded Back",1951.84,8,0.0,585.552
166,CA-2014-139892,2014-09-08,2014-09-12,Standard Class,BM-11140,Becky Martin,Consumer,United States,San Antonio,Texas,78207,Central,TEC-MA-10000822,Technology,Machines,Lexmark MX611dhe Monochrome Laser Printer,8159.952,8,0.4,-1359.992
168,CA-2014-139892,2014-09-08,2014-09-12,Standard Class,BM-11140,Becky Martin,Consumer,United States,San Antonio,Texas,78207,Central,FUR-CH-10004287,Furniture,Chairs,SAFCO Arco Folding Chair,1740.06,9,0.3,-24.858
216,CA-2015-146262,2015-01-02,2015-01-09,Standard Class,VW-21775,Victoria Wilson,Corporate,United States,Medina,Ohio,44256,East,TEC-MA-10000864,Technology,Machines,Cisco 9971 IP Video Phone Charcoal,1188.0,9,0.7,-950.4


- selecting records having `sales >1000`

In [0]:
select * from sample_superstore where `Order Date`>='2017-01-01' limit 10

Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
13,CA-2017-114412,2017-04-15,2017-04-20,Standard Class,AA-10480,Andrew Allen,Consumer,United States,Concord,North Carolina,28027,South,OFF-PA-10002365,Office Supplies,Paper,Xerox 1967,15.552,3,0.2,5.4432
24,US-2017-156909,2017-07-16,2017-07-18,Second Class,SF-20065,Sandra Flanagan,Consumer,United States,Philadelphia,Pennsylvania,19140,East,FUR-CH-10002774,Furniture,Chairs,"Global Deluxe Stacking Chair, Gray",71.372,2,0.3,-1.0196
35,CA-2017-107727,2017-10-19,2017-10-23,Second Class,MA-17560,Matt Abelman,Home Office,United States,Houston,Texas,77095,Central,OFF-PA-10000249,Office Supplies,Paper,Easy-staple paper,29.472,3,0.2,9.9468
42,CA-2017-120999,2017-09-10,2017-09-15,Standard Class,LC-16930,Linda Cazamias,Corporate,United States,Naperville,Illinois,60540,Central,TEC-PH-10004093,Technology,Phones,Panasonic Kx-TS550,147.168,4,0.2,16.5564
44,CA-2017-139619,2017-09-19,2017-09-23,Standard Class,ES-14080,Erin Smith,Corporate,United States,Melbourne,Florida,32935,South,OFF-ST-10003282,Office Supplies,Storage,"Advantus 10-Drawer Portable Organizer, Chrome Metal Frame, Smoke Drawers",95.616,2,0.2,9.5616
72,CA-2017-114440,2017-09-14,2017-09-17,Second Class,TB-21520,Tracy Blumstein,Consumer,United States,Jackson,Michigan,49201,Central,OFF-PA-10004675,Office Supplies,Paper,"Telephone Message Books with Fax/Mobile Section, 5 1/2"" x 3 3/16""",19.05,3,0.0,8.763
76,US-2017-118038,2017-12-09,2017-12-11,First Class,KB-16600,Ken Brennan,Corporate,United States,Houston,Texas,77041,Central,OFF-BI-10004182,Office Supplies,Binders,Economy Binders,1.248,3,0.8,-1.9344
77,US-2017-118038,2017-12-09,2017-12-11,First Class,KB-16600,Ken Brennan,Corporate,United States,Houston,Texas,77041,Central,FUR-FU-10000260,Furniture,Furnishings,"6"" Cubicle Wall Clock, Black",9.708,3,0.6,-5.8248
78,US-2017-118038,2017-12-09,2017-12-11,First Class,KB-16600,Ken Brennan,Corporate,United States,Houston,Texas,77041,Central,OFF-ST-10000615,Office Supplies,Storage,"SimpliFile Personal File, Black Granite, 15w x 6-15/16d x 11-1/4h",27.24,3,0.2,2.724
85,US-2017-119662,2017-11-13,2017-11-16,First Class,CS-12400,Christopher Schild,Home Office,United States,Chicago,Illinois,60623,Central,OFF-ST-10003656,Office Supplies,Storage,Safco Industrial Wire Shelving,230.376,3,0.2,-48.9549


- Displaying `Orders after 2017`

# **4. Aggregations**

### **Total Sales by Category**
- Finding which category generates maximum revenue.
- Using GROUP BY, SUM

In [0]:
select Category,round(sum(Sales),2) as tot_sales from sample_superstore group by Category limit 10;

Category,tot_sales
Furniture,741999.8
Office Supplies,719047.03
Technology,836154.03


### **Total Quantity Sold by Category**

In [0]:
select Category,sum(Quantity) as tot_sales from sample_superstore group by Category limit 10

Category,tot_sales
Furniture,8028
Office Supplies,22906
Technology,6939


### **Average Profit By Category**

In [0]:
select Category,round(avg(Profit),2) from sample_superstore group by Category limit 10;

Category,"round(avg(Profit),2)"
Furniture,8.7
Office Supplies,20.33
Technology,78.75


**Insight**
- Technology generally contributes high sales.

# **5.Sort and limit results (top products, top categories)**

### **Top 10 Products by Sales**
- Identify best-selling products.
- `ORDER BY`,` LIMIT`

In [0]:
select `Product Name`,round(sum(Sales)) as tot_sales ,count(*) as quantity from sample_superstore group by `Product Name` order by tot_sales desc limit 10;

Product Name,tot_sales,quantity
Canon imageCLASS 2200 Advanced Copier,61600.0,5
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind,27453.0,10
Cisco TelePresence System EX90 Videoconferencing Unit,22638.0,1
HON 5400 Series Task Chairs for Big and Tall,21871.0,8
GBC DocuBind TL300 Electric Binding System,19823.0,11
GBC Ibimaster 500 Manual ProClick Binding System,19025.0,9
Hewlett Packard LaserJet 3310 Copier,18840.0,8
"HP Designjet T520 Inkjet Large Format Printer - 24"" Color",18375.0,3
GBC DocuBind P400 Electric Binding System,17965.0,6
High Speed Automatic Electric Letter Opener,17030.0,3


### **Top Categories By Sales** 

In [0]:
select Category,round(sum(Sales),2) as tot_sales,count(*) as quantity from sample_superstore group by Category order by tot_sales desc limit 10;

Category,tot_sales,quantity
Technology,836154.03,1847
Furniture,741999.8,2121
Office Supplies,719047.03,6026


- The above gives us the insight about top revenue generating products and the categories

### **Seeing Total Profit alongside Sales**

In [0]:
select `product name`,round(sum(sales)) as tot_sales,round(sum(profit)) as tot_profit,count(*) as quantity from sample_superstore group by `product name` order by tot_sales desc limit 10;

product name,tot_sales,tot_profit,quantity
Canon imageCLASS 2200 Advanced Copier,61600.0,25200.0,5
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind,27453.0,7753.0,10
Cisco TelePresence System EX90 Videoconferencing Unit,22638.0,-1811.0,1
HON 5400 Series Task Chairs for Big and Tall,21871.0,0.0,8
GBC DocuBind TL300 Electric Binding System,19823.0,2234.0,11
GBC Ibimaster 500 Manual ProClick Binding System,19025.0,761.0,9
Hewlett Packard LaserJet 3310 Copier,18840.0,6984.0,8
"HP Designjet T520 Inkjet Large Format Printer - 24"" Color",18375.0,4095.0,3
GBC DocuBind P400 Electric Binding System,17965.0,-1878.0,6
High Speed Automatic Electric Letter Opener,17030.0,-262.0,3


### **seeing the Average Discount % given on these products**

In [0]:
select `product name`,round(sum(sales)) as tot_sales,round(avg(discount)*100,1) as avg_discount_percent,count(*) as quantity from sample_superstore group by `product name` order by tot_sales desc limit 10;

product name,tot_sales,avg_discount_percent,quantity
Canon imageCLASS 2200 Advanced Copier,61600.0,12.0,5
Fellowes PB500 Electric Punch Plastic Comb Binding Machine with Manual Bind,27453.0,24.0,10
Cisco TelePresence System EX90 Videoconferencing Unit,22638.0,50.0,1
HON 5400 Series Task Chairs for Big and Tall,21871.0,20.0,8
GBC DocuBind TL300 Electric Binding System,19823.0,30.0,11
GBC Ibimaster 500 Manual ProClick Binding System,19025.0,52.2,9
Hewlett Packard LaserJet 3310 Copier,18840.0,20.0,8
"HP Designjet T520 Inkjet Large Format Printer - 24"" Color",18375.0,16.7,3
GBC DocuBind P400 Electric Binding System,17965.0,45.0,6
High Speed Automatic Electric Letter Opener,17030.0,6.7,3


# **6.Solve use cases (monthly trends, top customers, duplicates)**

### **Monthly Trends (Sales and Orders over time)**
- Study growth over time.
- GROUP BY, Date Functions

In [0]:
select year(`Order Date`) as year,month(`Order Date`) as month,round(sum(sales),2) as total_sales from sample_superstore group by year(`Order Date`),month(`Order Date`) order by year,month limit 10;

year,month,total_sales
2014,1,14236.89
2014,2,4519.89
2014,3,55691.01
2014,4,28295.34
2014,5,23648.29
2014,6,34595.13
2014,7,33946.39
2014,8,27909.47
2014,9,81777.35
2014,10,31453.39


Insight
Sales patterns can be observed month-wise

### **Top 10 Customers**
- Find valuable customers.
- Topic: GROUP BY, ORDER BY

In [0]:
select `Customer Name`,round(sum(sales),2) as total_sales,count(*) as quantity from sample_superstore group by `Customer Name` order by total_sales desc limit 10;

Customer Name,total_sales,quantity
Sean Miller,25043.05,15
Tamara Chand,19052.22,12
Raymond Buch,15117.34,18
Tom Ashbrook,14595.62,10
Adrian Barton,14473.57,20
Ken Lonsdale,14175.23,29
Sanjit Chand,14142.33,22
Hunter Lopez,12873.3,11
Sanjit Engle,12209.44,19
Christopher Conant,12129.07,11


**Insight**
- Few customers contribute a large share of sales.

### **Finding Duplicate Orders**
- Check data quality.
- GROUP BY, HAVING

In [0]:
select `Order ID`,`Product ID`,count(*) as cnt from sample_superstore group by `Order ID`,`Product ID` having count(*)>1;

Order ID,Product ID,cnt
CA-2016-129714,OFF-PA-10001970,2
US-2016-123750,TEC-AC-10004659,2
CA-2016-137043,FUR-FU-10003664,2
CA-2017-152912,OFF-ST-10003208,2
US-2014-150119,FUR-CH-10002965,2
CA-2015-103135,OFF-BI-10000069,2
CA-2017-118017,TEC-AC-10002006,2
CA-2016-140571,OFF-PA-10001954,2


**Why I Grouped by Both IDs**
- One Order, Multiple Items: A customer can buy a chair and a pen. This creates two rows with the same Order ID but different Product IDs. This is normal.

- The Error: If the exact same Product ID appears twice under the same Order ID, the system accidentally double-counted the item.

- The Solution: Grouping by both IDs skips normal orders and only flags true data glitches.

### **Loss-Making `Orders`**

In [0]:
select `Order ID`,`Product ID`,round(sum(profit),2) as total_loss from sample_superstore group by `Order ID`,`Product ID` having sum(profit)<0 order by total_loss asc limit 10;

Order ID,Product ID,total_loss
CA-2016-108196,TEC-MA-10000418,-6599.98
US-2017-168116,TEC-MA-10004125,-3839.99
CA-2014-169019,OFF-BI-10004995,-3701.89
CA-2017-134845,TEC-MA-10000822,-3399.98
US-2017-122714,OFF-BI-10001120,-2929.48
CA-2015-147830,TEC-MA-10000418,-2639.99
CA-2017-131254,OFF-BI-10003527,-2287.78
CA-2015-116638,FUR-TA-10000198,-1862.31
CA-2016-130946,OFF-BI-10004995,-1850.95
CA-2014-145317,TEC-MA-10002412,-1811.08


- groups by both order_id and product_id to look at every single item line.
- calculates the total profit for that line and filters for anything less than 0 (having sum(profit)<0).
- sorts it so your biggest financial losses appear right at the top.

# **7. Validation and Data Quality**

### **Total Row Count**

In [0]:
select count(*) from sample_superstore;

count(*)
9994


### **Check Null Values**

In [0]:
select count(*)-count(`Order ID`) OrderID_Nulls,count(*)-count(`Customer ID`) CustomerID_Nulls,
count(*)-count(Sales) Sales_Nulls
from sample_superstore;

OrderID_Nulls,CustomerID_Nulls,Sales_Nulls
0,0,0


### **Negative Profits**

In [0]:
select * from sample_superstore where Profit<0 limit 10;

Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
4,US-2015-108966,2015-10-11,2015-10-18,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.031
15,US-2015-118983,2015-11-22,2015-11-26,Standard Class,HP-14815,Harold Pawlan,Home Office,United States,Fort Worth,Texas,76106,Central,OFF-AP-10002311,Office Supplies,Appliances,"Holmes Replacement Filter for HEPA Air Cleaner, Very Large Room, HEPA Filter",68.81,5,0.8,-123.858
16,US-2015-118983,2015-11-22,2015-11-26,Standard Class,HP-14815,Harold Pawlan,Home Office,United States,Fort Worth,Texas,76106,Central,OFF-BI-10000756,Office Supplies,Binders,Storex DuraTech Recycled Plastic Frosted Binders,2.544,3,0.8,-3.816
24,US-2017-156909,2017-07-16,2017-07-18,Second Class,SF-20065,Sandra Flanagan,Consumer,United States,Philadelphia,Pennsylvania,19140,East,FUR-CH-10002774,Furniture,Chairs,"Global Deluxe Stacking Chair, Gray",71.372,2,0.3,-1.0196
28,US-2015-150630,2015-09-17,2015-09-21,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,Pennsylvania,19140,East,FUR-BO-10004834,Furniture,Bookcases,"Riverside Palais Royal Lawyers Bookcase, Royale Cherry Finish",3083.43,7,0.5,-1665.0522
29,US-2015-150630,2015-09-17,2015-09-21,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,Pennsylvania,19140,East,OFF-BI-10000474,Office Supplies,Binders,Avery Recycled Flexi-View Covers for Binding Systems,9.618,2,0.7,-7.0532
33,US-2015-150630,2015-09-17,2015-09-21,Standard Class,TB-21520,Tracy Blumstein,Consumer,United States,Philadelphia,Pennsylvania,19140,East,OFF-BI-10001525,Office Supplies,Binders,"Acco Pressboard Covers with Storage Hooks, 14 7/8"" x 11"", Executive Red",6.858,6,0.7,-5.715
37,CA-2016-117590,2016-12-08,2016-12-10,First Class,GH-14485,Gene Hale,Corporate,United States,Richardson,Texas,75080,Central,FUR-FU-10003664,Furniture,Furnishings,"Electrix Architect's Clamp-On Swing Arm Lamp, Black",190.92,5,0.6,-147.963
39,CA-2015-117415,2015-12-27,2015-12-31,Standard Class,SN-20710,Steve Nguyen,Home Office,United States,Houston,Texas,77041,Central,FUR-BO-10002545,Furniture,Bookcases,"Atlantic Metals Mobile 3-Shelf Bookcases, Custom Colors",532.3992,3,0.32,-46.9764
40,CA-2015-117415,2015-12-27,2015-12-31,Standard Class,SN-20710,Steve Nguyen,Home Office,United States,Houston,Texas,77041,Central,FUR-CH-10004218,Furniture,Chairs,"Global Fabric Manager's Chair, Dark Gray",212.058,3,0.3,-15.147


### **Distinct Customers**

In [0]:
select count(distinct `Customer ID`) from sample_superstore;

count(DISTINCTCustomer ID)
793
